File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/NotificationChannelManager.java
/*
 * © 2016-2018 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.ui;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.os.Build;
import android.support.annotation.RequiresApi;

import com.flowcrypt.email.R;

/**
 * This manager does job of register {@link NotificationChannel} of the app. The {@link NotificationChannel} was
 * added in the {@link Build.VERSION_CODES#O} and doesn't work on previous Android versions.
 *
 * @author Denis Bondarenko
 *         Date: 17.10.2017
 *         Time: 12:12
 *         E-mail: DenBond7@gmail.com
 */

public class NotificationChannelManager {
    public static final String CHANNEL_ID_ATTACHMENTS = "Attachments";
    public static final String CHANNEL_ID_MESSAGES = "Messages";

    /**
     * Register {@link NotificationChannel}(s) of the app in the system.
     *
     * @param context Interface to global information about an application environment.
     */
    public static void registerNotificationChannels(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationManager notificationManager =
                    (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

            if (notificationManager != null) {
                notificationManager.createNotificationChannel(generateAttachmentsNotificationChannel(context));
                notificationManager.createNotificationChannel(generateMessagesNotificationChannel(context));
            }
        }
    }

    /**
     * Generate messages notification channel.
     *
     * @param context Interface to global information about an application environment.
     * @return Generated {@link NotificationChannel}
     */
    @RequiresApi(api = Build.VERSION_CODES.O)
    private static NotificationChannel generateMessagesNotificationChannel(Context context) {
        CharSequence name = context.getString(R.string.messages);
        String description = context.getString(R.string.messages_notification_channel);
        int importance = NotificationManager.IMPORTANCE_DEFAULT;

        NotificationChannel notificationChannel = new NotificationChannel(CHANNEL_ID_MESSAGES, name, importance);
        notificationChannel.setDescription(description);
        notificationChannel.enableLights(true);
        notificationChannel.enableVibration(true);

        return notificationChannel;
    }

    /**
     * Generate attachments notification channel.
     *
     * @param context Interface to global information about an application environment.
     * @return Generated {@link NotificationChannel}
     */
    @RequiresApi(api = Build.VERSION_CODES.O)
    private static NotificationChannel generateAttachmentsNotificationChannel(Context context) {
        CharSequence name = context.getString(R.string.attachments);
        String description = context.getString(R.string.download_attachments_notification_channel);
        int importance = NotificationManager.IMPORTANCE_DEFAULT;

        NotificationChannel notificationChannel = new NotificationChannel(CHANNEL_ID_ATTACHMENTS, name, importance);
        notificationChannel.setDescription(description);
        notificationChannel.enableLights(false);
        notificationChannel.enableVibration(false);

        return notificationChannel;
    }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/notifications/NotificationChannelManager.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.ui.notifications

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import com.flowcrypt.email.R

/**
 * This manager does job of register [NotificationChannel] of the app.
 *
 * @author Denys Bondarenko
 */
object NotificationChannelManager {
  const val CHANNEL_ID_ATTACHMENTS = "Attachments"
  const val CHANNEL_ID_MESSAGES = "Messages"
  const val CHANNEL_ID_SYSTEM = "System"
  const val CHANNEL_ID_ERRORS = "Errors"
  const val CHANNEL_ID_SYNC = "Sync"
  const val CHANNEL_ID_SILENT = "Silent"

  /**
   * Register [NotificationChannel](s) of the app in the system.
   *
   * @param context Interface to global information about an application environment.
   */
  fun registerNotificationChannels(context: Context) {
    val notificationManager =
      context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    notificationManager.createNotificationChannel(genGeneralNotificationChannel(context))
    notificationManager.createNotificationChannel(genAttsNotificationChannel(context))
    notificationManager.createNotificationChannel(genMsgsNotificationChannel(context))
    notificationManager.createNotificationChannel(genErrorNotificationChannel(context))
    notificationManager.createNotificationChannel(genSyncNotificationChannel(context))
    notificationManager.createNotificationChannel(genSilentNotificationChannel(context))
  }

  /**
   * Generate messages notification channel.
   *
   * @param context Interface to global information about an application environment.
   * @return Generated [NotificationChannel]
   */
  private fun genMsgsNotificationChannel(context: Context): NotificationChannel {
    val name = context.getString(R.string.messages)
    val description = context.getString(R.string.messages_notification_channel)
    val importance = NotificationManager.IMPORTANCE_DEFAULT

    val notificationChannel = NotificationChannel(CHANNEL_ID_MESSAGES, name, importance)
    notificationChannel.description = description
    notificationChannel.enableLights(true)
    notificationChannel.enableVibration(true)

    return notificationChannel
  }

  /**
   * Generate attachments notification channel.
   *
   * @param context Interface to global information about an application environment.
   * @return Generated [NotificationChannel]
   */
  private fun genAttsNotificationChannel(context: Context): NotificationChannel {
    val name = context.getString(R.string.attachments)
    val description = context.getString(R.string.download_attachments_notification_channel)
    val importance = NotificationManager.IMPORTANCE_DEFAULT

    val notificationChannel = NotificationChannel(CHANNEL_ID_ATTACHMENTS, name, importance)
    notificationChannel.description = description
    notificationChannel.enableLights(false)
    notificationChannel.enableVibration(false)

    return notificationChannel
  }

  /**
   * Generate general notification channel.
   *
   * @param context Interface to global information about an application environment.
   * @return Generated [NotificationChannel]
   */
  private fun genGeneralNotificationChannel(context: Context): NotificationChannel {
    val name = context.getString(R.string.system)
    val description = context.getString(R.string.system_notifications_notification_channel)
    val importance = NotificationManager.IMPORTANCE_DEFAULT

    val notificationChannel = NotificationChannel(CHANNEL_ID_SYSTEM, name, importance)
    notificationChannel.description = description
    notificationChannel.enableLights(true)
    notificationChannel.enableVibration(true)

    return notificationChannel
  }

  /**
   * Generate errors notification channel.
   *
   * @param context Interface to global information about an application environment.
   * @return Generated [NotificationChannel]
   */
  private fun genErrorNotificationChannel(context: Context): NotificationChannel {
    val name = context.getString(R.string.errors_notifications)
    val description = context.getString(R.string.errors_notifications_notification_channel)
    val importance = NotificationManager.IMPORTANCE_HIGH

    val notificationChannel = NotificationChannel(CHANNEL_ID_ERRORS, name, importance)
    notificationChannel.description = description
    notificationChannel.enableLights(true)
    notificationChannel.enableVibration(true)

    return notificationChannel
  }

  /**
   * Generate sync notification channel.
   *
   * @param context Interface to global information about an application environment.
   * @return Generated [NotificationChannel]
   */
  private fun genSyncNotificationChannel(context: Context): NotificationChannel {
    val name = context.getString(R.string.sync)
    val description = context.getString(R.string.sync_notifications_notification_channel)
    val importance = NotificationManager.IMPORTANCE_LOW

    val notificationChannel = NotificationChannel(CHANNEL_ID_SYNC, name, importance)
    notificationChannel.description = description
    notificationChannel.enableLights(false)
    notificationChannel.enableVibration(false)

    return notificationChannel
  }

  /**
   * Generate silent notification channel.
   *
   * @param context Interface to global information about an application environment.
   * @return Generated [NotificationChannel]
   */
  private fun genSilentNotificationChannel(context: Context): NotificationChannel {
    val name = context.getString(R.string.silent)
    val description = context.getString(R.string.silent_notifications_notification_channel)
    val importance = NotificationManager.IMPORTANCE_LOW

    val notificationChannel = NotificationChannel(CHANNEL_ID_SILENT, name, importance)
    notificationChannel.description = description

    return notificationChannel
  }
}
